keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterSpec.kt

```diff
@@ -0,0 +1,76 @@
+package com.netflix.spinnaker.keel.api.ec2
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.netflix.spinnaker.keel.api.Monikered
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.model.Moniker
+
+fun ClusterSpec.resolve(): Set<ServerGroupSpec> {
+  return locations.regions.map {
+    ServerGroupSpec(
+      moniker = moniker,
+      location = Location(
+        locations.accountName,
+        it.region,
+        it.subnet,
+        it.availabilityZones
+      ),
+      launchConfiguration = LaunchConfigurationSpec(
+        imageProvider = imageProvider,
+        instanceType = checkNotNull(overrides[it.region]?.launchConfiguration?.instanceType ?: defaults.launchConfiguration.instanceType),
+        ebsOptimized = checkNotNull(overrides[it.region]?.launchConfiguration?.ebsOptimized ?: defaults.launchConfiguration.ebsOptimized),
+        iamRole = checkNotNull(overrides[it.region]?.launchConfiguration?.iamRole ?: defaults.launchConfiguration.iamRole),
+        keyPair = checkNotNull(overrides[it.region]?.launchConfiguration?.keyPair ?: defaults.launchConfiguration.keyPair),
+        instanceMonitoring = overrides[it.region]?.launchConfiguration?.instanceMonitoring ?: defaults.launchConfiguration.instanceMonitoring ?: false,
+        ramdiskId = overrides[it.region]?.launchConfiguration?.ramdiskId ?: defaults.launchConfiguration.ramdiskId
+      ),
+      capacity = overrides[it.region]?.capacity ?: defaults.capacity ?: Capacity(1, 1, 1),
+      // TODO: need to be able to merge these rather than treating them as blocks
+      dependencies = overrides[it.region]?.dependencies ?: defaults.dependencies ?: Dependencies(),
+      health = overrides[it.region]?.health ?: defaults.health ?: Health(),
+      scaling = overrides[it.region]?.scaling ?: defaults.scaling ?: Scaling(),
+      tags = overrides[it.region]?.tags ?: defaults.tags ?: emptyMap()
+    )
+  }
+    .toSet()
+}
+
+data class ClusterSpec(
+  override val moniker: Moniker,
+  val imageProvider: ImageProvider,
+  val locations: ClusterLocations,
+  val defaults: ClusterServerGroupSpec,
+  val overrides: Map<String, ClusterServerGroupSpec> = emptyMap()
+) : Monikered, ResourceSpec {
+  @JsonIgnore
+  override val id = "${locations.accountName}:${moniker.name}"
+}
+
+data class ClusterLocations(
+  val accountName: String,
+  val regions: Set<ClusterRegion>
+)
+
+data class ClusterRegion(
+  val region: String,
+  val subnet: String,
```

Comment:
Another one: one single subnet, but multiple AZs? How does that map out?

Reply:
AFAIK subnets are region-wide. Most teams allow their server groups to balance across all available AZs but there are some teams that pin to particular AZs or even have a separate server group per AZ.

Reply:
Hmm... From the [AWS docs](https://docs.aws.amazon.com/vpc/latest/userguide/working-with-vpcs.html#AddaSubnet)
> Availability Zone: Optionally choose an Availability Zone in which your subnet will reside, or leave the default No Preference to let AWS choose an Availability Zone for you.

I totally understand balancing across AZs -- that's a recommended best practices. Which is why I wasn't sure what the single `subnet` property would do. Judging by the UI, it doesn't look like you manually create/specify subnets when creating a server group, only AZs. So where would this value come from?

Sorry for all the questions... 😅

Reply:
Questions are good! I _think_ (and you probably want to check with somebody who knows the AWS environment better than me) that the same logical subnet name (which is what this field contains) maps to different subnet ids in each AZ under the hood. Whatever the exact story is there, it's handled in CloudDriver.

Reply:
Ah! I think you're right:
![Screen Shot 2019-09-19 at 4 14 24 PM](https://user-images.githubusercontent.com/1323478/65287612-aa92c200-daf8-11e9-8c14-08b4cafc915f.png)

Reply:
@luispollo I think it's a valid question whether we can expect OSS users to have things set up that way